File: front/src/conversation/rendering/amrp.ts
import type {
  AgentTaskBlock,
  AIMessageData,
  AudioBlock,
  CitationSource,
  FileBlock,
  ImageBlock,
  MessageState,
  RichBlock,
  ToolBlock,
  VideoBlock,
} from "./types";

function firstString(source: Record<string, any>, keys: string[]): string {
  for (const key of keys) {
    const value = source[key];
    if (value !== undefined && value !== null && String(value).trim()) {
      return String(value).trim();
    }
  }
  return "";
}

function normalizeRole(role: unknown): AIMessageData["role"] {
  const value = String(role ?? "").toLowerCase();
  if (value === "user") return "user";
  if (value === "system") return "system";
  return "assistant";
}

export function normalizeMessageState(message: Record<string, any>): MessageState {
  const raw = String(message.state ?? message.status ?? "").toLowerCase();
  if (["queued", "pending"].includes(raw)) return "queued";
  if (["streaming", "sending", "generating", "collecting", "processing"].includes(raw)) {
    return "streaming";
  }
  if (["interrupted", "paused"].includes(raw)) return "interrupted";
  if (["failed", "error"].includes(raw)) return "failed";
  if (["cancelled", "canceled", "stopped"].includes(raw)) return "cancelled";
  if (message.typingStart && message.typingDone !== true) return "streaming";
  return "completed";
}

function normalizeStatus(value: unknown): ToolBlock["status"] {
  const raw = String(value ?? "").toLowerCase();
  if (["queued", "pending"].includes(raw)) return "queued";
  if (["running", "streaming", "sending"].includes(raw)) return "running";
  if (["failed", "error"].includes(raw)) return "failed";
  if (["cancelled", "canceled", "stopped"].includes(raw)) return "cancelled";
  return "success";
}

function parseJsonValue(value: unknown): unknown {
  if (typeof value !== "string") return value;
  const trimmed = value.trim();
  if (!trimmed || (!trimmed.startsWith("{") && !trimmed.startsWith("["))) return value;
  try {
    return JSON.parse(trimmed);
  } catch {
    return value;
  }
}

function normalizeBlock(raw: Record<string, any>, index: number): RichBlock {
  const type = String(raw.kind ?? raw.type ?? raw.blockType ?? "").toLowerCase();
  const id = firstString(raw, ["id", "blockId"]) || `block-${index}`;
  if (type.includes("tool")) {
    return {
      kind: "tool",
      id,
      name: firstString(raw, ["name", "toolName", "title"]) || "工具调用",
      arguments: raw.arguments ?? raw.args ?? raw.input,
      result: raw.result ?? raw.output ?? raw.content,
      duration: Number(raw.duration ?? raw.durationMs ?? 0) || undefined,
      error: firstString(raw, ["error", "errorCode", "errorMessage"]) || undefined,
      status: normalizeStatus(raw.status ?? raw.state),
    };
  }
  if (type.includes("file") || type.includes("attachment")) {
    return {
      kind: "file",
      id,
      name: firstString(raw, ["name", "fileName", "title"]) || "未命名文件",
      mimeType: firstString(raw, ["mimeType", "mime"]) || undefined,
      size: Number(raw.size ?? raw.sizeBytes ?? raw.fileSize ?? 0) || undefined,
      url: firstString(raw, ["url", "resourceUri", "href"]) || undefined,
      status: String(raw.status ?? "ready").toLowerCase() === "failed"
        ? "failed"
        : String(raw.status ?? "ready").toLowerCase() === "loading"
          ? "loading"
          : "ready",
      error: firstString(raw, ["error", "errorMessage"]) || undefined,
    };
  }
  if (type.includes("image")) {
    return {
      kind: "image",
      id,
      url: firstString(raw, ["url", "imageUrl", "src"]),
      alt: firstString(raw, ["alt", "altText", "title"]) || undefined,
      mimeType: firstString(raw, ["mimeType", "mime"]) || undefined,
      animated: Boolean(raw.animated ?? raw.isAnimated),
      width: Number(raw.width ?? 0) || undefined,
      height: Number(raw.height ?? 0) || undefined,
      status: "loading",
    };
  }
  if (type.includes("audio") || type.includes("voice")) {
    return {
      kind: "audio",
      id,
      url: firstString(raw, ["url", "audioUrl", "src"]),
      title: firstString(raw, ["title", "fileName", "name"]) || "语音",
      duration: Number(raw.duration ?? raw.durationMs ?? 0) || undefined,
      status: "loading",
    };
  }
  if (type.includes("video")) {
    return {
      kind: "video",
      id,
      url: firstString(raw, ["url", "videoUrl", "src"]),
      title: firstString(raw, ["title", "fileName", "name"]) || "视频",
      poster: firstString(raw, ["poster", "posterUrl"]) || undefined,
      duration: Number(raw.duration ?? raw.durationMs ?? 0) || undefined,
      status: "loading",
    };
  }
  if (type.includes("artifact")) {
    return {
      kind: "artifact",
      id,
      artifactKind: firstString(raw, ["artifactKind", "artifactType", "kindName"]) || "document",
      title: firstString(raw, ["title", "name"]) || "Artifact",
      mimeType: firstString(raw, ["mimeType", "mime"]) || undefined,
      content: typeof raw.content === "string" ? raw.content : undefined,
      fileId: firstString(raw, ["fileId"]) || undefined,
      url: firstString(raw, ["url", "href"]) || undefined,
      size: Number(raw.size ?? raw.sizeBytes ?? 0) || undefined,
    };
  }
  if (type.includes("agent") || type.includes("task") || type.includes("workflow")) {
    const steps = Array.isArray(raw.steps)
      ? raw.steps.map((step: any, stepIndex: number) => ({
          id: String(step?.id ?? `step-${stepIndex}`),
          title: firstString(step ?? {}, ["title", "name", "label"]) || `Step ${stepIndex + 1}`,
          status: normalizeStatus(step?.status ?? step?.state) === "success"
            ? "done" as const
            : normalizeStatus(step?.status ?? step?.state) === "failed"
              ? "failed" as const
              : normalizeStatus(step?.status ?? step?.state) === "running"
                ? "running" as const
                : "pending" as const,
          meta: firstString(step ?? {}, ["meta", "statusLabel", "duration"]) || undefined,
        }))
      : [];
    const status = normalizeStatus(raw.status ?? raw.state);
    return {
      kind: "agent-task",
      id,
      title: firstString(raw, ["title", "name", "taskTitle"]) || "Agent Task",
      status: status === "success" ? "done" : status === "failed" ? "failed" : status === "running" ? "running" : "pending",
      steps,
      progress: Number(raw.progress ?? raw.percentage ?? 0) || undefined,
      elapsed: firstString(raw, ["elapsed", "elapsedTime"]) || undefined,
      error: firstString(raw, ["error", "errorMessage"]) || undefined,
    } satisfies AgentTaskBlock;
  }
  if (type.includes("extension") || firstString(raw, ["rendererId"])) {
    return {
      kind: "extension",
      id,
      rendererId: firstString(raw, ["rendererId"]) || "unknown.renderer",
      version: Number(raw.version ?? 1),
      payload: raw.payload ?? raw.data,
    };
  }
  return {
    kind: "unknown",
    id,
    type: type || "unknown",
    payload: raw,
  };
}

function legacyBlocks(message: Record<string, any>): RichBlock[] {
  if (Array.isArray(message.blocks)) {
    return message.blocks
      .filter((item: unknown): item is Record<string, any> => !!item && typeof item === "object")
      .map(normalizeBlock);
  }
  const blocks: RichBlock[] = [];
  const type = String(message.msgType ?? message.messageType ?? message.type ?? "text").toLowerCase();
  const commonId = String(message.id ?? `legacy-${Date.now()}`);
  if (type === "tool_call" || type === "toolcall" || type === "tool") {
    blocks.push({
      kind: "tool",
      id: `${commonId}-tool`,
      name: firstString(message, ["toolName", "name", "title"]) || "工具调用",
      arguments: parseJsonValue(message.toolArguments ?? message.arguments),
      result: parseJsonValue(message.toolResult ?? message.result ?? message.content),
      duration: Number(message.duration ?? message.durationMs ?? 0) || undefined,
      error: firstString(message, ["error", "errorMessage"]) || undefined,
      status: normalizeStatus(message.status),
    });
  }
  if (type === "agent_task" || type === "agenttask" || type === "agent-task") {
    blocks.push(normalizeBlock({ ...message, kind: "agent-task" }, 0));
  }
  if (type === "file") {
    blocks.push({
      kind: "file",
      id: `${commonId}-file`,
      name: firstString(message, ["fileName", "name"]) || "未命名文件",
      mimeType: firstString(message, ["mimeType", "mime"]) || undefined,
      size: Number(message.fileSizeBytes ?? message.sizeBytes ?? message.fileSize ?? 0) || undefined,
      url: firstString(message, ["resourceUri", "url", "href"]) || undefined,
      status: String(message.status ?? "").toLowerCase() === "failed" ? "failed" : "ready",
      error: firstString(message, ["error", "errorMessage"]) || undefined,
    } satisfies FileBlock);
  }
  if (type === "image" || firstString(message, ["imageUrl"])) {
    const imageUrl = firstString(message, ["imageUrl", "image_url"]);
    if (imageUrl) {
      blocks.push({
        kind: "image",
        id: `${commonId}-image`,
        url: imageUrl,
        alt: firstString(message, ["altText", "alt"]) || undefined,
        mimeType: firstString(message, ["mimeType", "mime"]) || undefined,
        animated: Boolean(message.isAnimated ?? message.is_animated),
        width: Number(message.width ?? message.media_width ?? 0) || undefined,
        height: Number(message.height ?? message.media_height ?? 0) || undefined,
        status: "loading",
      } satisfies ImageBlock);
    }
  }
  if (type === "audio" || type === "voice" || firstString(message, ["audioUrl"])) {
    const audioUrl = firstString(message, ["audioUrl", "audio_url"]);
    if (audioUrl) {
      blocks.push({
        kind: "audio",
        id: `${commonId}-audio`,
        url: audioUrl,
        title: firstString(message, ["fileName", "name"]) || "语音",
        duration: Number(message.audioDuration ?? 0) || undefined,
        status: "loading",
      } satisfies AudioBlock);
    }
  }
  if (type === "video" || firstString(message, ["videoUrl"])) {
    const videoUrl = firstString(message, ["videoUrl", "video_url"]);
    if (videoUrl) {
      blocks.push({
        kind: "video",
        id: `${commonId}-video`,
        url: videoUrl,
        title: firstString(message, ["fileName", "name"]) || "视频",
        poster: firstString(message, ["poster", "posterUrl"]) || undefined,
        duration: Number(message.videoDuration ?? 0) || undefined,
        status: "loading",
      } satisfies VideoBlock);
    }
  }
  if (Array.isArray(message.attachments)) {
    for (const attachment of message.attachments) {
      if (attachment && typeof attachment === "object") {
        blocks.push(normalizeBlock(attachment, blocks.length));
      }
    }
  }
  return blocks;
}

function normalizeSources(message: Record<string, any>): CitationSource[] {
  const raw = message.sources ?? message.citations;
  if (!Array.isArray(raw)) return [];
  return raw
    .filter((item: unknown): item is Record<string, any> => !!item && typeof item === "object")
    .map((item, index) => ({
      id: firstString(item, ["id", "index"]) || String(index + 1),
      title: firstString(item, ["title", "name", "url", "fileId"]) || `Source ${index + 1}`,
      url: firstString(item, ["url", "href"]) || undefined,
      fileId: firstString(item, ["fileId"]) || undefined,
      snippet: firstString(item, ["snippet", "text", "description"]) || undefined,
    }));
}

export function normalizeAIMessage(
  input: Record<string, any>,
  character?: { id?: string; name?: string; avatar?: string },
): AIMessageData {
  const message = input ?? {};
  const role = normalizeRole(message.role);
  const type = String(message.msgType ?? message.messageType ?? message.type ?? "text").toLowerCase();
  let markdown = typeof message.content === "string" ? message.content : String(message.text ?? "");
  if ((type === "image" && markdown === "[图片]") || (type === "video" && markdown === "[视频]")) {
    markdown = "";
  }
  if (["tool_call", "toolcall", "tool", "agent_task", "agenttask", "agent-task"].includes(type)) {
    markdown = "";
  }
  const reasoningContent = firstString(message, ["reasoningContent", "reasoning_content"]);
  const rawThinking = message.thinking;
  const thinkingContent =
    rawThinking && typeof rawThinking === "object"
      ? firstString(rawThinking, ["content", "text"])
      : reasoningContent;
  const thinkingDuration = Number(rawThinking?.duration ?? message.reasoningDuration ?? 0);
  const reasoningDurationMs = Number(message.reasoningDurationMs ?? 0);
  const thinkingDurationSeconds =
    Number.isFinite(thinkingDuration) && thinkingDuration > 0
      ? thinkingDuration
      : Number.isFinite(reasoningDurationMs) && reasoningDurationMs > 0
        ? reasoningDurationMs / 1000
        : undefined;
  const state = normalizeMessageState(message);
  return {
    id: String(message.id ?? `message-${Date.now()}`),
    conversationId: firstString(message, ["conversationId", "conversation_id"]) || undefined,
    role,
    character: character
      ? {
          id: String(character.id ?? ""),
          name: String(character.name ?? "Amitia"),
          avatar: character.avatar,
        }
      : undefined,
    markdown,
    thinking:
      thinkingContent || message.generationPending === true
        ? {
            content: thinkingContent,
            state: message.generationPending === true ? "streaming" : state,
            duration: thinkingDurationSeconds,
          }
        : undefined,
    blocks: legacyBlocks(message),
    sources: normalizeSources(message),
    state,
    createdAt: Date.parse(String(message.createdAt ?? "")) || Number(message.createdAt) || Date.now(),
    raw: message,
  };
}

export function richBlockText(block: RichBlock): string {
  switch (block.kind) {
    case "tool":
      return [block.name, formatUnknown(block.arguments), formatUnknown(block.result), block.error]
        .filter(Boolean)
        .join("\n");
    case "file":
      return [block.name, block.mimeType, block.url].filter(Boolean).join("\n");
    case "image":
      return block.alt || block.url;
    case "audio":
    case "video":
      return [block.title, block.url].filter(Boolean).join("\n");
    case "artifact":
      return [block.title, block.artifactKind, block.content].filter(Boolean).join("\n");
    case "agent-task":
      return [block.title, ...block.steps.map((step) => `${step.title} ${step.meta ?? ""}`)].join("\n");
    case "extension":
      return `${block.rendererId}\n${formatUnknown(block.payload)}`;
    case "unknown":
      return formatUnknown(block.payload);
  }
}

export function formatUnknown(value: unknown): string {
  if (value === undefined || value === null) return "";
  if (typeof value === "string") return value;
  try {
    return JSON.stringify(value, null, 2);
  } catch {
    return String(value);
  }
}

export function aimMessagePlainText(message: AIMessageData): string {
  return [message.markdown, message.thinking?.content, ...message.blocks.map(richBlockText)]
    .filter((value) => String(value ?? "").trim())
    .join("\n\n");
}

File: mobile_app/lib/features/conversation/rendering/amrp.dart
import '../../../shared/models/models.dart';

enum AmrpMessageRole { user, assistant, system }

enum AmrpMessageState {
  queued,
  streaming,
  completed,
  interrupted,
  failed,
  cancelled,
}

enum AmrpToolStatus { queued, running, success, failed, cancelled }

enum AmrpAssetStatus { loading, ready, failed }

enum AmrpAgentStepStatus { pending, running, done, failed }

class AmrpCharacter {
  final String id;
  final String name;
  final String avatar;

  const AmrpCharacter({
    this.id = '',
    this.name = 'Amitia',
    this.avatar = '',
  });
}

class AmrpThinkingBlock {
  final String content;
  final AmrpMessageState state;
  final Duration? duration;

  const AmrpThinkingBlock({
    required this.content,
    this.state = AmrpMessageState.completed,
    this.duration,
  });
}

class AmrpCitationSource {
  final String id;
  final String title;
  final String url;
  final String fileId;
  final String snippet;

  const AmrpCitationSource({
    required this.id,
    required this.title,
    this.url = '',
    this.fileId = '',
    this.snippet = '',
  });
}

sealed class AmrpRichBlock {
  final String id;

  const AmrpRichBlock({required this.id});
}

class AmrpToolBlock extends AmrpRichBlock {
  final String name;
  final Object? arguments;
  final Object? result;
  final Duration? duration;
  final String error;
  final AmrpToolStatus status;

  const AmrpToolBlock({
    required super.id,
    required this.name,
    this.arguments,
    this.result,
    this.duration,
    this.error = '',
    this.status = AmrpToolStatus.success,
  });
}

class AmrpFileBlock extends AmrpRichBlock {
  final String name;
  final String mimeType;
  final int size;
  final String url;
  final AmrpAssetStatus status;
  final String error;

  const AmrpFileBlock({
    required super.id,
    required this.name,
    this.mimeType = '',
    this.size = 0,
    this.url = '',
    this.status = AmrpAssetStatus.ready,
    this.error = '',
  });
}

class AmrpImageBlock extends AmrpRichBlock {
  final String url;
  final String alt;
  final String mimeType;
  final bool animated;
  final double width;
  final double height;
  final AmrpAssetStatus status;

  const AmrpImageBlock({
    required super.id,
    required this.url,
    this.alt = '',
    this.mimeType = '',
    this.animated = false,
    this.width = 0,
    this.height = 0,
    this.status = AmrpAssetStatus.loading,
  });
}

class AmrpAudioBlock extends AmrpRichBlock {
  final String url;
  final String title;
  final Duration? duration;
  final AmrpAssetStatus status;

  const AmrpAudioBlock({
    required super.id,
    required this.url,
    this.title = '语音',
    this.duration,
    this.status = AmrpAssetStatus.loading,
  });
}

class AmrpVideoBlock extends AmrpRichBlock {
  final String url;
  final String title;
  final String poster;
  final Duration? duration;
  final AmrpAssetStatus status;

  const AmrpVideoBlock({
    required super.id,
    required this.url,
    this.title = '视频',
    this.poster = '',
    this.duration,
    this.status = AmrpAssetStatus.loading,
  });
}

class AmrpArtifactBlock extends AmrpRichBlock {
  final String artifactKind;
  final String title;
  final String mimeType;
  final String content;
  final String fileId;
  final String url;
  final int size;

  const AmrpArtifactBlock({
    required super.id,
    required this.title,
    this.artifactKind = 'document',
    this.mimeType = '',
    this.content = '',
    this.fileId = '',
    this.url = '',
    this.size = 0,
  });
}

class AmrpAgentStep {
  final String id;
  final String title;
  final AmrpAgentStepStatus status;
  final String meta;

  const AmrpAgentStep({
    required this.id,
    required this.title,
    this.status = AmrpAgentStepStatus.pending,
    this.meta = '',
  });
}

class AmrpAgentTaskBlock extends AmrpRichBlock {
  final String title;
  final AmrpAgentStepStatus status;
  final List<AmrpAgentStep> steps;
  final int progress;
  final String elapsed;
  final String error;

  const AmrpAgentTaskBlock({
    required super.id,
    required this.title,
    this.status = AmrpAgentStepStatus.pending,
    this.steps = const <AmrpAgentStep>[],
    this.progress = 0,
    this.elapsed = '',
    this.error = '',
  });
}

class AmrpExtensionBlock extends AmrpRichBlock {
  final String rendererId;
  final int version;
  final Object? payload;

  const AmrpExtensionBlock({
    required super.id,
    required this.rendererId,
    this.version = 1,
    this.payload,
  });
}

class AmrpUnknownBlock extends AmrpRichBlock {
  final String type;
  final Object? payload;

  const AmrpUnknownBlock({
    required super.id,
    required this.type,
    this.payload,
  });
}

class AmrpMessage {
  final String id;
  final String conversationId;
  final AmrpMessageRole role;
  final AmrpCharacter character;
  final String markdown;
  final AmrpThinkingBlock? thinking;
  final List<AmrpRichBlock> blocks;
  final List<AmrpCitationSource> sources;
  final AmrpMessageState state;
  final DateTime createdAt;
  final ChatMessage raw;

  const AmrpMessage({
    required this.id,
    required this.conversationId,
    required this.role,
    required this.character,
    required this.markdown,
    required this.blocks,
    required this.sources,
    required this.state,
    required this.createdAt,
    required this.raw,
    this.thinking,
  });

  static AmrpMessage fromChatMessage(
    ChatMessage message, {
    AmrpCharacter character = const AmrpCharacter(),
  }) {
    return AmrpMessage(
      id: message.id,
      conversationId: '',
      role: switch (message.role) {
        MessageRole.user => AmrpMessageRole.user,
        MessageRole.system => AmrpMessageRole.system,
        MessageRole.assistant => AmrpMessageRole.assistant,
      },
      character: character,
      markdown: _markdownFor(message),
      thinking: message.reasoningContent.trim().isEmpty
          ? null
          : AmrpThinkingBlock(
              content: message.reasoningContent,
              state: _stateFor(message),
            ),
      blocks: _blocksFor(message),
      sources: const <AmrpCitationSource>[],
      state: _stateFor(message),
      createdAt: message.time,
      raw: message,
    );
  }

  static AmrpMessageState _stateFor(ChatMessage message) {
    return switch (message.status) {
      MessageStatus.sending => AmrpMessageState.streaming,
      MessageStatus.error => AmrpMessageState.failed,
      MessageStatus.sent || MessageStatus.delivered => AmrpMessageState.completed,
    };
  }

  static String _markdownFor(ChatMessage message) {
    if (message.type == MessageType.toolCall ||
        message.type == MessageType.agentTask ||
        message.type == MessageType.systemNotice) {
      return '';
    }
    if (message.type == MessageType.image && message.content == '[图片]') {
      return '';
    }
    if (message.type == MessageType.video && message.content == '[视频]') {
      return '';
    }
    if (message.type == MessageType.audio && message.content == '[语音]') {
      return '';
    }
    return message.content;
  }

  static List<AmrpRichBlock> _blocksFor(ChatMessage message) {
    final blocks = <AmrpRichBlock>[];
    switch (message.type) {
      case MessageType.toolCall:
        blocks.add(
          AmrpToolBlock(
            id: '${message.id}:tool',
            name: message.toolName?.trim().isNotEmpty == true
                ? message.toolName!.trim()
                : '工具调用',
            result: message.toolResult,
            error: message.status == MessageStatus.error
                ? (message.toolResult ?? '')
                : '',
            status: message.status == MessageStatus.error
                ? AmrpToolStatus.failed
                : message.status == MessageStatus.sending
                ? AmrpToolStatus.running
                : AmrpToolStatus.success,
          ),
        );
        break;
      case MessageType.agentTask:
        final steps = (message.agentTaskSteps ?? const <String>[])
            .asMap()
            .entries
            .map(
              (entry) => AmrpAgentStep(
                id: '${message.id}:step:${entry.key}',
                title: entry.value,
                status: _agentStatusFromProgress(
                  entry.key,
                  message.agentTaskProgress ?? 0,
                ),
              ),
            )
            .toList(growable: false);
        blocks.add(
          AmrpAgentTaskBlock(
            id: '${message.id}:agent',
            title: message.agentTaskTitle?.trim().isNotEmpty == true
                ? message.agentTaskTitle!.trim()
                : 'Agent Task',
            status: message.status == MessageStatus.error
                ? AmrpAgentStepStatus.failed
                : message.status == MessageStatus.sending
                ? AmrpAgentStepStatus.running
                : AmrpAgentStepStatus.done,
            steps: steps,
            progress: message.agentTaskProgress ?? 0,
            elapsed: message.agentTaskElapsed ?? '',
          ),
        );
        break;
      case MessageType.file:
        blocks.add(
          AmrpFileBlock(
            id: '${message.id}:file',
            name: message.fileName?.trim().isNotEmpty == true
                ? message.fileName!.trim()
                : '未命名文件',
            mimeType: message.mimeType ?? '',
            size: (message.fileSizeKB ?? 0) * 1024,
            url: message.resourceUri ?? '',
            status: message.status == MessageStatus.error
                ? AmrpAssetStatus.failed
                : AmrpAssetStatus.ready,
            error: message.status == MessageStatus.error ? message.content : '',
          ),
        );
        break;
      case MessageType.image:
        final url = (message.mediaUrl ?? message.resourceUri ?? '').trim();
        if (url.isNotEmpty) {
          blocks.add(
            AmrpImageBlock(
              id: '${message.id}:image',
              url: url,
              alt: message.fileName ?? '',
              mimeType: message.mimeType ?? '',
            ),
          );
        }
        break;
      case MessageType.audio:
        final url = (message.mediaUrl ?? message.resourceUri ?? '').trim();
        if (url.isNotEmpty) {
          blocks.add(
            AmrpAudioBlock(
              id: '${message.id}:audio',
              url: url,
              title: message.fileName ?? '语音',
              duration: message.durationMs == null
                  ? null
                  : Duration(milliseconds: message.durationMs!),
            ),
          );
        }
        break;
      case MessageType.video:
        final url = (message.mediaUrl ?? message.resourceUri ?? '').trim();
        if (url.isNotEmpty) {
          blocks.add(
            AmrpVideoBlock(
              id: '${message.id}:video',
              url: url,
              title: message.fileName ?? '视频',
              duration: message.durationMs == null
                  ? null
                  : Duration(milliseconds: message.durationMs!),
            ),
          );
        }
        break;
      case MessageType.text:
      case MessageType.code:
      case MessageType.emote:
      case MessageType.systemNotice:
        break;
    }
    return blocks;
  }

  static AmrpAgentStepStatus _agentStatusFromProgress(int index, int progress) {
    if (progress <= 0) return AmrpAgentStepStatus.pending;
    final completedSteps = (progress / 25).floor();
    if (index < completedSteps) return AmrpAgentStepStatus.done;
    if (index == completedSteps) return AmrpAgentStepStatus.running;
    return AmrpAgentStepStatus.pending;
  }

  String get plainText {
    final values = <String>[
      markdown,
      if (thinking != null) thinking!.content,
      ...blocks.map(_blockText),
    ];
    return values.where((value) => value.trim().isNotEmpty).join('\n\n');
  }

  static String _blockText(AmrpRichBlock block) {
    return switch (block) {
      AmrpToolBlock() =>
        '${block.name}\n${_stringify(block.arguments)}\n${_stringify(block.result)}',
      AmrpFileBlock() => '${block.name} ${block.mimeType} ${block.url}',
      AmrpImageBlock() => block.alt.isNotEmpty ? block.alt : block.url,
      AmrpAudioBlock() => '${block.title} ${block.url}',
      AmrpVideoBlock() => '${block.title} ${block.url}',
      AmrpArtifactBlock() => '${block.title}\n${block.content}',
      AmrpAgentTaskBlock() =>
        '${block.title}\n${block.steps.map((step) => step.title).join('\n')}',
      AmrpExtensionBlock() => '${block.rendererId}\n${_stringify(block.payload)}',
      AmrpUnknownBlock() => _stringify(block.payload),
    };
  }

  static String _stringify(Object? value) {
    if (value == null) return '';
    if (value is String) return value;
    return value.toString();
  }
}

